src/supernova/mod.rs

```diff
@@ -129,139 +113,226 @@ where
 {
   fn write_bytes<W: Sized + io::Write>(&self, byte_sink: &mut W) -> Result<(), io::Error> {
-    assert!(!self.claims.is_empty());
-    for claim in &self.claims {
-      claim.get_public_params().write_bytes(byte_sink)?;
+    assert!(!self.circuit_params.is_empty());
+    for pp in &self.circuit_params {
+      pp.write_bytes(byte_sink)?;
     }
     Ok(())
   }
 }
 
-impl<G1, G2, C1, C2> RunningClaims<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
+  C1: EnforcingStepCircuit<G1::Scalar>,
+  C2: EnforcingStepCircuit<G2::Scalar>,
 {
-  fn new(mut claims: Vec<RunningClaim<G1, G2, C1, C2>>) -> Self {
-    let running_claim_params = claims
+  /// new running claim params
+  pub fn new<NC: NonUniformCircuit<G1, G2, C1, C2>>(
+    non_unifrom_circuit: &NC,
+  ) -> Self {
+    // uhhh if this aint zero this doesn't work
+    let _initial_pc = non_unifrom_circuit.initial_program_counter();
+    let num_circuits = non_unifrom_circuit.num_circuits();
+
+    let c_primarys = (0..num_circuits)
+      .map(|i| non_unifrom_circuit.primary_circuit(i))
+      .collect::<Vec<_>>();
+    let c_secondary = non_unifrom_circuit.secondary_circuit();
+
+    let circuit_params = c_primarys
       .iter()
-      .map(|c| c.get_public_params())
+      .map(|c_primary| CircuitParams::setup_without_commitkey(c_primary, num_circuits))
       .collect::<Vec<_>>();
+    let ck_primary = Self::compute_primary_ck(&circuit_params);
 
-    let (ck_primary, ck_secondary) = compute_commitment_keys(&running_claim_params);
+    let augmented_circuit_params_secondary =
+      SuperNovaAugmentedCircuitParams::new(BN_LIMB_WIDTH, BN_N_LIMBS, false);
+    let ro_consts_secondary: ROConstants<G2> = ROConstants::<G2>::default();
+    let F_arity_secondary = c_secondary.arity();
+    let ro_consts_circuit_secondary: ROConstantsCircuit<G1> = ROConstantsCircuit::<G1>::default();
 
-    for claim in &mut claims {
-      claim.set_commitment_key(ck_primary.clone(), ck_secondary.clone());
-    }
-    RunningClaims {
-      claims,
+    let circuit_secondary: SuperNovaAugmentedCircuit<'_, G1, C2> = SuperNovaAugmentedCircuit::new(
+      &augmented_circuit_params_secondary,
+      None,
+      &c_secondary,
+      ro_consts_circuit_secondary.clone(),
+      num_circuits,
+    );
+    let mut cs: ShapeCS<G2> = ShapeCS::new();
+    let _ = circuit_secondary.synthesize(&mut cs);
+    let (r1cs_shape_secondary, ck_secondary) = cs.r1cs_shape_and_key(None);
+
+    let pp = PublicParams {
+      circuit_params,
+      ck_primary,
+      F_arity_secondary,
+      ro_consts_secondary,
+      ro_consts_circuit_secondary,
+      ck_secondary,
+      r1cs_shape_secondary,
+      augmented_circuit_params_secondary,
       digest: OnceCell::new(),
+      _p: PhantomData,
+    };
+
+    pp.digest(); // make sure to compute the digest
```

Comment:
Please clarify why in the comment. We wouldn't like folks to cargo-cult this call.